Tests/AuthFoundationTests/APIRetryTests.swift

```diff
@@ -0,0 +1,119 @@
+        XCTAssertEqual(client.request?.allHTTPHeaderFields?["X-Okta-Retry-For"], requestId)
+    }
+    
+    func testCustomRetryCount() throws {
+        client = MockApiClient(configuration: configuration,
+                                   session: urlSession,
+                                   baseURL: baseUrl,
+                                   shouldRetry: .retry(maximumCount: 5))
+        try performRetryRequest(count: 6)
+        XCTAssertEqual(client.request?.allHTTPHeaderFields?["X-Okta-Retry-Count"], "5")
+        XCTAssertEqual(client.request?.allHTTPHeaderFields?["X-Okta-Retry-For"], requestId)
+    }
+    
+    func testApiRateLimit() throws {
+        let date = "Fri, 09 Sep 2022 02:22:14 GMT"
+        let rateLimit = APIRateLimit(with: ["x-rate-limit-limit": "0",
+                                            "x-rate-limit-remaining": "0",
+                                            "x-rate-limit-reset": "1662690193",
+                                            "Date": date])
+        XCTAssertNotNil(rateLimit?.delay)
+        XCTAssertEqual(rateLimit?.delay, 59.0)
+    }
+    
+    func performRetryRequest(count: Int, isSuccess: Bool = false) throws {
+        let date = Date()
+        for _ in 0..<count {
+            urlSession.expect("https://example.okta.com/oauth2/v1/token",
+                              data: try data(from: .module, for: "token", in: "MockResponses"),
+                              statusCode: 429,
+                              contentType: "application/json",
+                              headerFields: ["x-rate-limit-limit": "0",
+                                             "x-rate-limit-remaining": "0",
+                                             "x-rate-limit-reset": "1609459200",
+                                             "Date": "\(date)",
+                                             "x-okta-request-id": requestId])
+        }
+        
+        if isSuccess {
+            urlSession.expect("https://example.okta.com/oauth2/v1/token",
+                              data: try data(from: .module, for: "token", in: "MockResponses"),
+                              contentType: "application/json",
+                              headerFields: ["Date": "\(date)",
+                                             "x-okta-request-id": requestId])
+        }
+        
+        let expect = expectation(description: "network request")
+        apiRequest.send(to: client, completion: { result in
+            switch result {
+            case .success(let response):
+                XCTAssertNotNil(response)
+                XCTAssertEqual(response.requestId, self.requestId)
+            case .failure(let error):
+                XCTAssertEqual(error.errorDescription, APIClientError.statusCode(429).errorDescription)
+            }
+            expect.fulfill()
+        })
+        waitForExpectations(timeout: 2.0) { error in
+            XCTAssertNil(error)
+        }
+    }
+}
```

Comment:
Could you also add tests with malformed or missing Date and X-Rate-Limit-Reset headers? Current behavior throws an error in that case, but I think it makes sense to capture that as part of a test.

Reply:
I have included test for missing X-Rate-Limit-Reset headers and not missing date. we internally default to Date.nowCoordinated if the date header is missing.